config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java

```diff
@@ -73,67 +84,134 @@ public DeploymentSpec read(Reader reader) {
+
+            for (Element topLevelTag : XML.getChildren(root)) {
+                if (topLevelTag.getTagName().equals(instanceTag))
+                    steps.addAll(readInstanceContent(topLevelTag.getAttribute(idAttribute), topLevelTag, new MutableOptional<>(), root));
+                else
+                    steps.addAll(readNonInstanceSteps(topLevelTag, new MutableOptional<>(), topLevelTag)); // (No global service id here)
             }
+        }
+
+        return new DeploymentSpec(steps,
+                                  optionalIntegerAttribute(majorVersionTag, root),
+                                  xmlForm);
+    }
+
+    /**
+     * Reads the content of an (implicit or explicit) instance tag producing an instances step
+     *
+     * @param instanceNameString a comma-separated list of the names of the instances this is for
+     * @param instanceTag the element having the content of this instance
+     * @param parentTag the parent of instanceTag (or the same, if this instances is implicitly defined which means instanceTag is the root)
+     * @return the instances specified, one for each instance name element
+     */
+    private List<DeploymentInstanceSpec> readInstanceContent(String instanceNameString,
+                                                             Element instanceTag,
+                                                             MutableOptional<String> globalServiceId,
+                                                             Element parentTag) {
+        if (validate)
+            validateTagOrder(instanceTag);
+
+        // Values where the parent may provide a default
+        DeploymentSpec.UpgradePolicy upgradePolicy = readUpgradePolicy(instanceTag, parentTag);
+        List<DeploymentSpec.ChangeBlocker> changeBlockers = readChangeBlockers(instanceTag, parentTag);
+        Optional<AthenzDomain> athenzDomain = stringAttribute(athenzDomainAttribute, instanceTag)
+                                                        .or(() -> stringAttribute(athenzDomainAttribute, parentTag))
+                                                        .map(AthenzDomain::from);
+        Optional<AthenzService> athenzService = stringAttribute(athenzServiceAttribute, instanceTag)
+                                                        .or(() -> stringAttribute(athenzServiceAttribute, parentTag))
+                                                        .map(AthenzService::from);
+        Notifications notifications = readNotifications(instanceTag, parentTag);
+
+        // Values where there is no default
+        List<Step> steps = new ArrayList<>();
+        for (Element instanceChild : XML.getChildren(instanceTag))
+            steps.addAll(readNonInstanceSteps(instanceChild, globalServiceId, instanceChild));
+        List<Endpoint> endpoints = readEndpoints(instanceTag);
+
+        // Build and return instances with these values
+        return Arrays.stream(instanceNameString.split("'"))
```

Comment:
Should be split by commas, no?

Reply:
Excellent catch, thanks! And this shows a test is missing ..